### ISSUE TITLE
.Net5.0 not respecting StaticWebAssetBasePath on publish of Blazor WASM project

### ISSUE THREAD

**Issue by chris03** [dotnet/aspnetcore#27776]
### Describe the bug
After upgrading to .Net5.0  I noticed that when publishing (`dotnet publish`) a Blazor WASM project the **StaticWebAssetBasePath** is not taken into account.
The files inside `wwwroot` are not put in the specified base folder.
However, when the project is run, the framework files are correctly located.

I'm following this documentation about `<StaticWebAssetBasePath>`
https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-5.0#hosted-deployment-with-multiple-blazor-webassembly-apps

### To Reproduce
Run:  `dotnet new blazorwasm`
Open the `.csproj` file
Add `<StaticWebAssetBasePath>test</StaticWebAssetBasePath>` inside `<PropertyGroup>`
Save & Run: `dotnet publish`
Look at the content of: `bin\Debug\net5.0\publish\wwwroot`

**Current result**: Most files are incorrectly located in `wwwroot` and some files are inside `wwwroot\test`
**Expected result**: All files (framework,css,...) should be inside `wwwroot\test`
![image](https://user-images.githubusercontent.com/1290874/98964028-53a73380-24d6-11eb-823e-bb182a8a508f.png)

I get the expected result when running the same steps with .NetCore 3.1 (`dotnet new blazorwasm -f netcoreapp3.1`) 
![image](https://user-images.githubusercontent.com/1290874/98964394-bef10580-24d6-11eb-9079-227e810f695b.png)

### Further technical details
- .Net 5.0
```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100
 Commit:    5044b93829

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100\

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

.NET SDKs installed:
  5.0.100 [C:\Program Files\dotnet\sdk]
```
I ran the commands on the command line.

**Comment by javiercn** [comment]
@chris03 thanks for the details.

@pranavkm this seems like a bug, but we do have tests for this, so I'm a bit confused. Can it be that we don't have a test that changes the static web asset base path?

**Comment by mvromer** [comment]
I'm seeing this as well. In my case I set `<StaticWebAssetBasePath>static</StaticWebAssetBasePath>` in my csproj file.

After setting a custom StaticWebAssetBasePath property, I'm also seeing the Blazor app fail to load when running `dotnet run`. In my case, this is what I tried:

1. Run `dotnet new blazorwasm -o ExampleClassifierApp`
2. Update ExampleClassifierApp.csproj with the new StaticWebAssetBasePath above.
3. Modify index.html and FetchData.razor so that paths to the static assets (CSS, JS, and JSON files) they request are prefixed with `static/`.

After starting `dotnet run` and going to https://localhost:5001, the app fails to load with the following errors in my dev tools:

![image](https://user-images.githubusercontent.com/1542776/99098956-5f126180-259f-11eb-88a7-fa77130eb302.png)

For grins I created a `wwwroot/static` folder inside my project's root directory and moved the static `css` and `sample-data` folders into that. Rerunning that shows that the static CSS loads now, but the blazor.boot.json can't be loaded:

![image](https://user-images.githubusercontent.com/1542776/99096942-83207380-259c-11eb-8100-4ea3cf80025a.png)

Looks like blazor.webassembly.js tries to fetch `_framework/blazor.boot.json`, which will be relative to where the index is being served from (the root in this case) and not relative to the path set for StaticWebAssetBasePath.

**Comment by alexchiraples** [comment]
I'm seeing this as well.

Context:
1. Blazor WebAssembly App, PWA, Core hosted, auth w/ Individual user accounts, created with 3.1 (app1)
2. Upgraded app1.Client, app1.Server and app1.Shared to net 5
3. Create a second app (app2) hosted, and configure it to connect to app1 server 
4. Configure app1 and app2 as Hosted deployment with multiple Blazor WebAssembly apps as documented https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-5.0#hosted-deployment-with-multiple-blazor-webassembly-apps

Goal:
Hosted deployment with multiple Blazor WebAssembly apps
- /app1
- /app2

After following the docs and receiving lots of errors, including those described in this issue, I've managed to bypass most of the errors. I'm stuck at a 401 error when calling the APIs even if the user is authenticated. For example: oidc.user:https://localhost:44352:app1 and app2 have all the required values. I am not sure if this this is related to the StaticWebAssetBasePath thing or if it should be another issue opened on this topic. 

Please find below sample of my configuration for achieving:
- hosted deployments
- multiple blazor webassembly apps
- different base path
- authentication (Identity)

** dotnet version **
```
λ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100
 Commit:    5044b93829

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100\

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

.NET SDKs installed:
  3.1.402 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Server/Startup.cs**
```
app.UseBlazorFrameworkFiles("/app1");
app.UseBlazorFrameworkFiles("/app2");
app.UseStaticFiles();

app.UseRouting();

app.UseIdentityServer();
app.UseAuthentication();
app.UseAuthorization();


app.UseEndpoints(endpoints =>
{
	endpoints.MapRazorPages();
	endpoints.MapControllers();

	endpoints.MapFallbackToFile("/app1/index.html");
	endpoints.MapFallbackToFile("/app1/{*path:nonfile}", "app1/index.html");
	endpoints.MapFallbackToFile("/app2/{*path:nonfile}", "app2/index.html");

	
	// default path to app1
	endpoints.MapGet("/", context => Task.Factory.StartNew(() => context.Response.Redirect("/app1", true, true)));

	endpoints.MapGet("/app1/_configuration/app1.Client", context => Task.Factory.StartNew(() => context.Response.Redirect("/_configuration/app1.Client", true, true)));
	endpoints.MapGet("/app2/_configuration/app2.Client", context => Task.Factory.StartNew(() => context.Response.Redirect("/_configuration/app2.Client", true, true)));

});
``` 
**Server/application.json**
```
"IdentityServer": {
        "Clients": {
            "app1.Client": {
                "Profile": "IdentityServerSPA",
                "RedirectUri": "/app1/authentication/login-callback",
                "LogoutUri": "/app1/authentication/logout-callback"
            },
            "app2.Client": {
                "Profile": "IdentityServerSPA",
                "RedirectUri": "/app2/authentication/login-callback",
                "LogoutUri": "/app2/authentication/logout-callback"
            }
        },
        "Key": {
            "Type": "Development"
        }
    }
```
**app1/wwwroot/index.html** (similar configuration for app2)
```
<head>
  ...
  <base href="/app1/" />
  ...
</head>
<body>
  ...
  <script src="/_content/Microsoft.AspNetCore.Components.WebAssembly.Authentication/AuthenticationService.js"></script>
  <script src="_framework/blazor.webassembly.js"></script>
  ...
</body>
```

**app1/Program.cs**
```
 var builder = WebAssemblyHostBuilder.CreateDefault(args);
builder.RootComponents.Add<App>("#app");


builder.Services.AddHttpClient("app1.ServerAPI", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
                .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();

// Supply HttpClient instances that include access tokens when making requests to the server project
builder.Services.AddScoped(sp => sp.GetRequiredService<IHttpClientFactory>().CreateClient("app1.ServerAPI"));

builder.Services.AddApiAuthorization();

await builder.Build().RunAsync();
```

After doing all of this, everything works except calling any API that requires authorization. #19928 looks similar.

**Comment by glararan** [comment]
I can confirm this bug aswell.. After upgrading from 3.1 to 5.0 publish is not respecting anymore base path and update wwwroot/_framework/blazor.boot.json to last project in solution.

**Comment by dazinator** [comment]
What's the best mechanism to be notified when 5.0.1 is released?

**Comment by achmstein** [comment]
I cannot make it work with IdentityServer! Is there any example for hosting multiple blazor apps with IdentityServer?

**Comment by boukenka** [comment]
@dazinator 5.0.1 is out! 😄

**Comment by lovedeep91** [comment]
T have already updated to 5.0.1 but same problem, now in iis pubblish folder in wwwroot i have Firstapp, Firstapp__framework, Firstapp_content, ecc

**Comment by javiercn** [comment]
@lovedeep91 thanks for contacting us.

We are aware of this issue, we realized that there was a small bug in our fix and we have another fix in place and automation to make sure this is fixed in 5.0.2.

**Comment by JonTvermose** [comment]
@lovedeep91 Same issue here. :-(

It looks like you can "fix" this by appending "/" to your <StaticWebAssetBasePath>, which in turn means it does not work when running the solution in Visual Studio.

So our workload now untill 5.0.2 seems to be to append "/" to the <StaticWebAssetBasePath> when we need to deploy, and then remove it again after deployment.

**Comment by dazinator** [comment]
> We are aware of this issue, we realized that there was a small bug in our fix and we have another fix in place and automation to make sure this is fixed in 5.0.2.

Is 5.0.2 going to be pushed out quickly, or are we waiting for next months release (assuming approx 1 servicing release a month) - I ask because having waited over a month already, this issue prevents us upgrading from 3.1 to 5.0, and for such a small issue, its a shame to be forced to wait another month over something that should have been fixed in 5.0.1
Also - should this issue be reopened / assigned to 5.0.2 milestone?

**Comment by javiercn** [comment]
@dazinator unfortunately we only do patches every month, we don't publish "hotfixes" on demand unless there is a super critical security issue. We do have an associated PR [here](https://github.com/dotnet/aspnetcore/pull/28539) that will be merged either today or tomorrow, but unfortunately our patch schedule is monthly.

If you **REALLY, REALLY** want to get unblocked now, you can apply the same changes that the PR includes in the SDK files within your dotnet installation **AT YOUR OWN RISK**, but an official release won't happen until sometime January.

**Comment by dazinator** [comment]
@javiercn 
> If you REALLY, REALLY want to get unblocked now, you can apply the same changes that the PR includes in the SDK files within your dotnet installation AT YOUR OWN RISK, but an official release won't happen until sometime January.

I wish I could, but our CI / CD pipeline is done via an Azure DevOps Pipelines "hosted" agent - which has the official dotnet sdk 5.0 installed. I'm sure there would be too much work involved trying to patch that at the start of every build, for it to be a cost effective endeavour. I think we are forced to wait, so wait we shall!

**Comment by javiercn** [comment]
@dazinator an alternative could be to leverage the configuration to configure different staticwebassetbasepaths

```csharp
<PropertyGroup>
  <StaticWebAssetBasePath>app</StaticWebAssetBasePath>
  <StaticWebAssetBasePath Condition="'$(Configuration)' == 'Release'>$(StaticWebAssetBasePath)/</StaticWebAssetBasePath>
</PropertyGroup>
```

Assuming run = debug, release = publish

**Comment by JonTvermose** [comment]
@javiercn 
Great - the files are in the correct folder. 👍 

However when I use the proposed solution I get the following error in the console:

```
Error parsing 'integrity' attribute ('undefined'). The hash algorithm must be one of 'sha256', 'sha384', or 'sha512', followed by a '-' character.

Uncaught (in promise) Error: Content hash is required
    at e.<anonymous> (blazor.webassembly.js:1)
    at blazor.webassembly.js:1
    at Object.next (blazor.webassembly.js:1)
    at blazor.webassembly.js:1
    at new Promise (<anonymous>)
    at r (blazor.webassembly.js:1)
    at e.loadResourceWithCaching (blazor.webassembly.js:1)
    at e.loadResource (blazor.webassembly.js:1)
    at blazor.webassembly.js:1
    at l (blazor.webassembly.js:1)
```

The blazor.boot.json file looks like this;

```
{
  "cacheBootResources": true,
  "config": [ ],
  "debugBuild": false,
  "entryAssembly": "Almego.Web.Blazor.Reports",
  "icuDataMode": 0,
  "linkerEnabled": true,
  "resources": {
    "assembly": { },
    "lazyAssembly": null,
    "pdb": null,
    "runtime": { },
    "satelliteResources": null
  }
}
```

Any ideas how we can solve this? Our deployment pipeline is currently blocked and we would very much like to not wait a full month before deploying again :-(

**Comment by javiercn** [comment]
Ah, seems like it'll be more problematic than expected since that value gets embedded in multiple places.

I think this will have to wait until 5.0.2 then.

**Comment by dazinator** [comment]
@javiercn - does the test case put in place to cover the fix in 5.0.2, cover this scenario that @JonTvermose  is seeing? We want to avoid waiting for 5.0.3 only to then have to wait again for a 5.0.4! :-)

**Comment by javiercn** [comment]
@dazinator the cases that I added should cover things E2E.

While we can't guarantee we didn't miss anything, we tried our best to make sure we fixed it for good.

**Comment by michaelpaulus** [comment]
This is the workaround I'm using until it is fixed, in case it helps anybody:
adding this to the project file, replace app_ with what your prefix is:

<Target Name="MoveBlazorFrameworkFiles" AfterTargets="Publish">
    <ItemGroup>
      <_BlazorFrameworkFiles Include="$(PublishDir)\wwwroot\app_framework\**\*.*" />
    </ItemGroup>
    <Move SourceFiles="@(_BlazorFrameworkFiles)" DestinationFolder="$(PublishDir)\wwwroot\app\_framework\%(RecursiveDir)" />
    <RemoveDir Directories="$(PublishDir)\wwwroot\app_framework" />
  </Target>

**Comment by fuzl-llc** [comment]
Thanks @michaelpaulus for your workaround. I was experiencing this issue as described in issue #29179 but it seems this thread covers my problem.  I have prefix "app1" and when I use your workaround with my prefix, doing "dotnet publish -c Release" gets the _framework files in the right place.  However, when I run it with this:

dotnet MyApp.Server.dll --urls http://localhost:5001

After blazor.webassembly.js, blazor.boot.json, etc. load properly, I get a request to this:

http://localhost:5001/app1/_framework/undefined

In Chrome dev tools I can see the request comes from initiator blazor.webassembly.js:1 and of course the request to this "undefined" URL returns a 404 and the Blazor app hangs. Has anyone else seen this behavior with the above fix?

**Comment by JonTvermose** [comment]
@fuzl-llc Just wanted to say that I spend days trying to fix this problem back in december with no luck. In the end we re-did our Blazor applications and consolidated them all under a single Blazor Webassembly application as we -needed- a fix (our production deployement was blocked due to this error).
We utilized the new render-modes introduced in 5.0.0 described here: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.rendering.rendermode?view=aspnetcore-5.0

I am watching #29179 as well, as we would ideally like to go back to having multiple Blazor apps in the same solution.

**Comment by fuzl-llc** [comment]
Thanks @JonTvermose for the info and link. I'm also blocked on a production deployment but my app is brand new and I only have a few users so I'm thinking of trying the potential workaround I mention over on #29179 in spite of it seeming risky and ill-advised...  I just don't want to turn back now.  Best of luck.  Hopefully MS will come through with a supported work-around soon or at the very least get this knocked out for good in 5.0.3.

**Comment by dazinator** [comment]
Im very concerned to he

**Comment by lovedeep91** [comment]
I have already updated to new version.. folders are now in correct position but like @JonTvermose i have same error now.

**Comment by dazinator** [comment]
Sadly I am not getting as far the publish completing any more. I'll raise a seperate issue.

**Comment by mikes-gh** [comment]
Is this still an issue.  I just spent hours on this.  Publishing to azure app service I need a trailing slash `<StaticWebassets>`
Then I need to change it back for testing
